### ISSUE TITLE
Excessive CPU usage

### ISSUE THREAD

**Issue by johnniesong** [wekan/wekan-mongodb#2]
https://github.com/wekan/wekan/issues/718 this issue also happened on your docker ENV.

Downgrading from Node 6.10.0 to v4.8.1 solved the 100% CPU issue

**Comment by xet7** [comment]
Fixed with downgrade to v4.8.1.

**Comment by mfshiu** [comment]
My Wekan node is v4.8.4 , has the same issue, CPU will be 100% usage when over 20 clients.
Should I downgrade to v4.8.1 ?

**Comment by xet7** [comment]
@mfshiu 

You can try, please add comment did it help.

**Comment by mfshiu** [comment]
It's not worked. 
Wekan make CPU loading very heavy when over 20 clients, always 100% CPU usage.

**Comment by xet7** [comment]
@mfshiu 

Ok, I think you are referring to this forum post:
https://discourse.wekan.io/t/cpu-utilization-problems-with-large-userbase/579/5

Wekan needs some profiling to identify bottlenecks etc.

**Comment by xet7** [comment]
@GavinLilly

Did you do any profiling of Wekan with Kadira? 

It could be possible that Meteor code tries to provide updates to all clients too fast, so some added delays in fetching and updating would made CPU load lower.

**Comment by xet7** [comment]
@GavinLilly

If MongoDB is not a bottleneck as you said, then having search #552 should be possible just with MongoDB.

**Comment by xet7** [comment]
@mfshiu 

I added to [Roadmap](https://github.com/wekan/wekan/wiki/Roadmap) Not started yet section "Try to fix Excessive CPU usage to encourage mfshiu to contribute his Wekan features in progress".

**Comment by mfshiu** [comment]
Thank you for your encourage, I will try.

**Comment by GavinLilly** [comment]
I started setting up Kadira but didn't get round to profiling. Other things took priority. I will be able to look at it in mid-December. I agree with your hypothesis in that perhaps Meteor might be trying to do too much to update to each client.
For the time being I've doubled the number of running Wekan instances to 6 across 3 dual core VMs. It seems more stable but the instances are still reaching 100% load.

On the MongoDB side I've not observed any increased CPU or RAM load so some form of search could be possible. I'd be a little concerned about it searching using a cursor instead of an index though

**Comment by xet7** [comment]
@GavinLilly 

Yes mid-December is great, thanks a lot!

I did also add note to search items issue that search should be done by using index.

**Comment by mfshiu** [comment]
I found the solution. The problem is "Frequent Subscriptions" in top of client\components\main\layouts.js
`
Meteor.subscribe('boards');
Meteor.subscribe('setting');
Meteor.subscribe('user-admin');
`
I move these code to top of client\components\boards\boardList.js
`
BlazeComponent.extendComponent({
    onCreated() {
        console.log("{boardList.js} (onCreated) begin");
        Meteor.subscribe('boards');
        Meteor.subscribe('setting');
        Meteor.subscribe('user-admin');
    },
`
CPU is no longer too high, from 100% down to 30% - 50%.
I will try to update the code or you can told me how to do that.

**Comment by mfshiu** [comment]
It seems I have no right to push the updates.
`remote: Permission to wekan/wekan.git denied to mfshiu.'
'fatal: unable to access 'https://github.com/wekan/wekan.git/': The requested URL returned error: 403`

**Comment by xet7** [comment]
@mfshiu 

You can submit pull request to wekan/wekan devel branch.

https://help.github.com/articles/fork-a-repo/

https://help.github.com/articles/creating-a-pull-request-from-a-fork/

https://help.github.com/articles/creating-a-pull-request/

**Comment by mfshiu** [comment]
Use this package is in Wekan VERY important: 

> [mixmax:smart-disconnect](https://atmospherejs.com/mixmax/smart-disconnect)

Meteor connection cost is very expensive, if your MongoDB contains large data, your active users are over several hundreds, you must use the package to auto disconnect user when they leave their focus on Wekan over 1 minute. I read [the article by Eliezer Steinbock](https://medium.freecodecamp.org/scaling-meteor-a-year-on-26ee37588e4b), then make my Wekan CPU ultra down.

**Comment by GavinLilly** [comment]
Is it possible to integrate smart-disconnect into the published release of Wekan after downloading or would I need to pull the source code and integrate before build/deploy?

**Comment by xet7** [comment]
@GavinLilly 

Yes, I will try to include smart-disconnect with Wekan.

**Comment by GavinLilly** [comment]
I'm not convinced that smart-disconnect was correctly bundled with the v0.62 release:

```
$ cat programs/server/packages/mixmax_smart-disconnect.js
(function () {



/* Exports */
if (typeof Package === 'undefined') Package = {};
Package['mixmax:smart-disconnect'] = {};

})();
```

As you can see no logic is included in the file. Is there another way to check if a package was included correctly?

**Comment by xet7** [comment]
@GavinLilly 

I only added package:
https://github.com/wekan/wekan/blob/devel/.meteor/versions#L95
https://github.com/wekan/wekan/blob/devel/.meteor/packages#L83

I don't know what else @mfshiu did to get it working.

At package page there is only info about IronRouter, not FlowRouter that Wekan uses:
https://atmospherejs.com/mixmax/smart-disconnect

If somebody knows how to get this working, please add pull request.

**Comment by GavinLilly** [comment]
Actually even though there doesn't seem to be any logic in that JS file I have observed that when I switch to a Wekan tab in Chrome I briefly see the message near the top of the page saying "Wekan is loading..." so it seems the smart disconnect package is doing it's job after all.

In addition, performance in our Wekan platform appears to be improved recently though that could also be affected by decreased traffic over the Christmas break. I'll report back if we experience further issues as traffic ramps up

**Comment by xet7** [comment]
@GavinLilly

Yes I noticed that behaviour also with Firefox previously, so smart disconnect seems to work.

**Comment by xet7** [comment]
Adding Redis Oplog would improve scalability:
- https://github.com/cult-of-coders/redis-oplog
- https://forums.meteor.com/t/meteor-scaling-redis-oplog-status-prod-ready/30855/479

**Comment by xet7** [comment]
Links for some future optimizations research:
- DTrace talks http://dtrace.org/blogs/bmc/2018/02/03/talks/ , HN comments at https://news.ycombinator.com/item?id=16303595

**Comment by GavinLilly** [comment]
We've fixed our CPU usage problem today with an environment change around Wekan. I wasn't aware during implementation that if you're using more than 1 instance of Wekan (or any MeteorJS based tool) you're supposed to set MONGO_OPLOG_URL as an environment variable. Without setting it, Meteor will perform a pull-and-diff update of it's dataset. With it, Meteor will update from the OPLOG. See [here](https://blog.meteor.com/tuning-meteor-mongo-livedata-for-scalability-13fe9deb8908)

After setting `MONGO_OPLOG_URL=mongodb://<username>:<password>@<mongoDbURL>/local?authSource=admin&replicaSet=rsWekan` the CPU usage for all Wekan instances dropped to an average of less than 10% with only occasional spikes to high usage (I guess when someone is doing a lot of work)

So as you can see it's not a problem with Wekan but how it's implemented in a given infrastructure. Apologies for the wild goose chase in previous months. Perhaps there should be an entry on the Wiki so users know to set this if they're using more than 1 instance of Wekan. Also perhaps it could be added to the docker compose yaml?

**Comment by xet7** [comment]
@GavinLilly 

What Wekan, Node and MongoDB version you use?

**Comment by xet7** [comment]
@GavinLilly 

What kind of infra you use? Number of instances, Docker, etc?

**Comment by xet7** [comment]
Nevermind, thanks for info, I'll add it :)

**Comment by xet7** [comment]
@kentonv

Currently Wekan get's complaints about 100% CPU usage when there is thousands of users on powerful servers, with this kind of setup at AWS:
https://github.com/wekan/wekan/wiki/AWS

Wekan uses Meteor 1.6.0.1, but most likely same problem is at 1.6.1.

There is currently these GitHub issues going on about this here:
https://github.com/meteor/meteor/issues/9796

Referencing to this issue you had:
https://github.com/meteor/meteor/issues/7747

Do you have ideas how this could be fixed in Wekan or Meteor ?

**Comment by kentonv** [comment]
@xet7 I commented on the other thread. You can try my patched node (linked there) and see if it helps. (If you don't want to compile yourself, copy `/opt/sandstorm/latest/bin/node` from your Sandstorm install.)

**Comment by xet7** [comment]
Please test newest Wekan v1.39. There has been speedups in Wekan v1.37 and v1.38. CPU usage also seems to be lower.
https://github.com/wekan/wekan/blob/devel/CHANGELOG.md
